src/triggers.h

```diff
@@ -1,7 +1,22 @@
 #pragma once
 
+#if defined(__arm__)
+// Force the compiler to generate A32 instructions for the two trigger functions
+// to ensure they end up being 4-byte (UPROBE_SWBP_INSN_SIZE) aligned, otherwise
+// uprobe_register() fails with EINVAL (see kernel/events/uprobes.c).
+#define __target_attr __attribute__((target("arm")))
```

Comment:
Thanks for reporting this, @daniloegea! It does look like a gcc bug; gcc 14 on [godbolt](https://gcc.godbolt.org/z/83TKac9h3) shows this:
```
<source>: In function 'int main()':
<source>:4:18: internal compiler error: canonical types differ for identical types 'void()' and 'void()'
    4 | int main() { call(f); }
      |              ~~~~^~~
```

Unfortunately, `aligned(4)` isn't sufficient -- we have to instruct the compiler to generate ARM code for these two trigger functions for uprobes to work:

- Kernel installs the probe by inserting a trap instruction ([`UPROBE_SWBP_ARM_INSN`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm/include/asm/uprobes.h?h=v5.10.182#n17)). This is `udf 25` in `arch/arm`, which has a different encoding in ARM vs Thumb (`0xe7f001f9` that the kernel uses is the A32 version). With `aligned(4)`, trigger functions are still Thumb code, so calling them won't cause the CPU to switch to ARM state.
- Similarly, even if we were to force the state change (e.g. by clearing the low-order bit of the function pointer), we'd have a state mismatch after returning from the uprobe (since `bx lr` in `{BEGIN,END}_trigger` is encoded as Thumb -- `0x4770` vs `0xe12fff1e`).

I'll play around with it some more to see if there's a workaround for gcc.

Reply:
@daniloegea: that might be the best solution, yeah.

This gcc issue is quite puzzling. Adding an unrelated function definition after the `target("arm")` one in your example causes it to compile fine: https://gcc.godbolt.org/z/hcoxEMs4e